cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/Operator.java

```diff
@@ -81,6 +82,12 @@ default void reconcileThese(String trigger, Set<NamespaceAndName> desiredNames,
             metrics().pausedResourceCounter(namespace).set(0);
         }
 
+        // Removing all CA cluster metrics not in the desired set
+        Set<String> desiredClusterNames = desiredNames.stream().map(NamespaceAndName::getName).collect(Collectors.toSet());
+        metrics().removeMetricsForCertificates(key -> key.getNamespace().equals(namespace)
```

Comment:
Quite sure, but last test is in progress :) Maciej will be back on tuesday and we will finalize.
But as I can see in code, it will be `*` in the `key` and thus it will work.

Reply:
> But as I can see in code, it will be * in the key and thus it will work.

You cannot have `*` in the key. The operator managing all namespaces - i.e. watching the namespace `*` - can have Kafka cluster named `my-cluster` in namespace A and another Kafka cluster named `my-cluster` in namespace B. And if the key has `*` as a namespace, then those will likely override each other, or?

Reply:
Will make sure to test it throughly on Tuesday 😄

Reply:
Ok, let me know on Tuesday. I think it now looks mostly good for me apart from this. Hopefully, @ppatierno can have a look soon as well so that we can proceed and merge this soon.

Reply:
I have made it even more explicit with namespace `*` in `removeMetrics`

Reply:
But you have to make sure that:
* You remove the metric for the cluster that was deleted
* Don't remove the metric for the cluster that was not deleted

So how will this work with clusters namespace-a/my-cluster and namespace-b/my-cluster? The code will delete both of them.

I think the only solution is to:
* Make sure the key always contains the namespace of the Kafka cluster and never the _watched namespace_ that might be `*`
* Make sure the metric is removed only when the name and namespace are both equal to what is in the key.

Reply:
But doesn't `*` in namespace means that it can be whatever and we clean it all?

Reply:
No. The `*` namespace is used to tell the assembly operators to watch the whole cluster instead of a single namespace. But each custom resource has its own namespace. The `*` is basically just an indicator of how should the watch (eventually some `list` requests be done).

I don't think the code you have now solves it. Maybe something like this:
```java
    public void removeMetrics(Set<NamespaceAndName> desiredNames, String namespace) {
        metrics.removeMetricsForCertificates(key -> {
            return desiredNames.contains(new NamespaceAndName(key.getNamespace(), key.getClusterName()));
        });
    }
```

I guess it might be useful to have some unit tests for the various combinations to see that the removal works well.